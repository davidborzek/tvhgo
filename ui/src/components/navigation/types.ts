import { ReactElement } from 'react';

export type INavigationItem = {
  to: string;
  title: string;
  requiredRoles?: string[];
  icon?: ReactElement;
  items?: INavigationItem[];
};
