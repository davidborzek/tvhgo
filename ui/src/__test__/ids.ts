export enum TestIds {
  PAGINATION_FIRST_PAGE = 'pagination-first-page-btn',
  PAGINATION_PREVIOUS_PAGE = 'pagination-previous-page-btn',
  PAGINATION_NEXT_PAGE = 'pagination-next-page-btn',
  PAGINATION_LAST_PAGE = 'pagination-last-page-btn',
  REVOKE_SESSION_BUTTON = 'revoke-session-btn',
  REVOKE_TOKEN_BUTTON = 'revoke-token-btn',
  THEME_DROPDOWN = 'theme-dropdown',
  LANGUAGE_DROPDOWN = 'language-dropdown',
  LOGOUT_BUTTON = 'logout-button',
  SAVE_USER_BUTTON = 'save-user-btn',
}
