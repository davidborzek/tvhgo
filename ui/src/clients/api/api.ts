import axios from 'axios';
import {
  Channel,
  CreateTokenResponse,
  EpgChannel,
  EpgEvent,
  ErrorResponse,
  ListResponse,
  Recording,
  Session,
  Token,
  TwoFactorAuthSettings,
  TwoFactorAuthSetupResult,
  UpdateRecording,
  UpdateUser,
  UpdateUserPassword,
  UserResponse,
} from './api.types';
import qs from 'qs';

type PaginationQuery = {
  limit?: number;
  offset?: number;
};

type SortQuery = {
  sort_key?: string;
  sort_dir?: string;
};

type PaginationSortQuery = PaginationQuery & SortQuery;

export type GetEpgEventsQuery = PaginationSortQuery & {
  title?: string;
  lang?: string;
  channel?: string;
  contentType?: string;
  fullText?: boolean;
  nowPlaying?: boolean;
  durationMin?: number;
  durationMax?: number;
  startsAt?: number;
  endsAt?: number;
};

export type GetEpgChannelEventsQuery = PaginationSortQuery & {
  startsAt?: number;
  endsAt?: number;
};

export type RecordingStatus = 'upcoming' | 'finished' | 'failed' | 'removed';

export type GetRecordingsQuery = PaginationSortQuery & {
  status?: RecordingStatus;
};

export type GetChannelsQuery = PaginationSortQuery & {
  name?: string;
};

export class ApiError extends Error {
  constructor(
    public readonly code: number,
    message: string
  ) {
    super(message);
  }
}

const client = axios.create({
  baseURL: '/api',
  paramsSerializer: (params) => qs.stringify(params, { arrayFormat: 'repeat' }),
});

client.interceptors.response.use(
  (res) => res,
  (err) => {
    if (axios.isAxiosError(err) && err.response) {
      const data = err.response.data as ErrorResponse;
      return Promise.reject(new ApiError(err.response.status, data.message));
    }
    return Promise.reject(err);
  }
);

export async function login(username: string, password: string, code?: string) {
  await client.post('/login', {
    username,
    password,
    totp: code,
  });
}

export async function logout(): Promise<void> {
  await client.post('/logout');
}

export async function getUser(): Promise<UserResponse> {
  const response = await client.get<UserResponse>('/user');
  return response.data;
}

export async function getEpgEvents(
  q?: GetEpgEventsQuery
): Promise<ListResponse<EpgEvent>> {
  const response = await client.get<ListResponse<EpgEvent>>('/epg/events', {
    params: q,
  });
  return response.data;
}

export async function getRelatedEpgEvents(
  id: number,
  q?: PaginationSortQuery
): Promise<ListResponse<EpgEvent>> {
  const response = await client.get<ListResponse<EpgEvent>>(
    `/epg/events/${id}/related`,
    {
      params: q,
    }
  );
  return response.data;
}

export async function getEpgEvent(id: number): Promise<EpgEvent> {
  const response = await client.get<EpgEvent>(`/epg/events/${id}`);
  return response.data;
}

export async function getEpg(
  q?: GetEpgChannelEventsQuery
): Promise<Array<EpgChannel>> {
  const response = await client.get<Array<EpgChannel>>('/epg', {
    params: q,
  });
  return response.data;
}

export async function getChannel(id: string): Promise<Channel> {
  const response = await client.get<Channel>(`/channels/${id}`);
  return response.data;
}

export async function getChannels(
  q?: GetChannelsQuery
): Promise<Array<Channel>> {
  const response = await client.get<Array<Channel>>(`/channels`, {
    params: q,
  });
  return response.data;
}

export async function fetch<T>(path: string): Promise<T> {
  const response = await client.get<T>(path);
  return response.data;
}

export async function recordByEvent(
  eventId: number,
  configId?: string
): Promise<void> {
  await client.post('/recordings/event', {
    eventId,
    configId,
  });
}

export async function stopRecording(id: string): Promise<void> {
  await client.put(`/recordings/${id}/stop`);
}

export async function cancelRecording(id: string): Promise<void> {
  await client.put(`/recordings/${id}/cancel`);
}

export async function removeRecording(id: string): Promise<void> {
  await client.delete(`/recordings/${id}`);
}

export async function stopRecordings(ids: string[]): Promise<void> {
  await client.put(`/recordings/stop`, null, {
    params: { ids },
  });
}

export async function cancelRecordings(ids: string[]): Promise<void> {
  await client.put(`/recordings/cancel`, null, {
    params: { ids },
  });
}

export async function removeRecordings(ids: string[]): Promise<void> {
  await client.delete(`/recordings`, {
    params: { ids },
  });
}

export async function getRecordings(
  q?: GetRecordingsQuery
): Promise<ListResponse<Recording>> {
  const response = await client.get<ListResponse<Recording>>(`/recordings`, {
    params: q,
  });
  return response.data;
}

export async function getRecording(id: string): Promise<Recording> {
  const response = await client.get<Recording>(`/recordings/${id}`);
  return response.data;
}

export async function updateRecording(
  id: string,
  opts: UpdateRecording
): Promise<void> {
  await client.patch(`/recordings/${id}`, opts);
}

export async function updateUser(opts: UpdateUser): Promise<UserResponse> {
  const response = await client.patch<UserResponse>(`/user`, opts);
  return response.data;
}

export async function updateUserPassword(
  opts: UpdateUserPassword
): Promise<void> {
  await client.patch(`/user/password`, opts);
}

export function getRecordingUrl(id: string): string {
  return `/api/recordings/${id}/stream`;
}

export async function getSessions(): Promise<Array<Session>> {
  const response = await client.get<Array<Session>>(`/sessions`);
  return response.data;
}

export async function deleteSession(id: number): Promise<void> {
  return await client.delete(`/sessions/${id}`);
}

export async function getTwoFactorAuthSettings(): Promise<TwoFactorAuthSettings> {
  const response = await client.get<TwoFactorAuthSettings>(`/two-factor-auth`);
  return response.data;
}

export async function deactivateTwoFactorAuth(code: string): Promise<void> {
  await client.put(`/two-factor-auth/deactivate`, {
    code,
  });
}

export async function setupTwoFactorAuth(
  password: string
): Promise<TwoFactorAuthSetupResult> {
  const response = await client.put<TwoFactorAuthSetupResult>(
    `/two-factor-auth/setup`,
    {
      password,
    }
  );
  return response.data;
}

export async function activateTwoFactorAuth(
  password: string,
  code: string
): Promise<void> {
  await client.put(`/two-factor-auth/activate`, {
    password,
    code,
  });
}

export async function getTokens(): Promise<Array<Token>> {
  const response = await client.get<Array<Token>>(`/tokens`);
  return response.data;
}

export async function deleteToken(id: number): Promise<void> {
  await client.delete(`/tokens/${id}`);
}

export async function createToken(name: string): Promise<CreateTokenResponse> {
  const response = await client.post<CreateTokenResponse>(`/tokens`, { name });
  return response.data;
}
