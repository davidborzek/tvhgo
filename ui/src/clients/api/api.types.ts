export type ListResponse<T> = {
  entries: T[];
  total: number;
  offset: number;
};

export type UserResponse = {
  id: string;
  username: string;
  email: string;
  displayNAme: string;
  createdAt: number;
  updatedAt: number;
};

export type ErrorResponse = {
  message: string;
};

export type EpgEvent = {
  id: number;
  audioDesc: boolean;
  channelId: string;
  channelName: string;
  channelNumber: number;
  piconId: number;
  description: string;
  endsAt: number;
  hd: boolean;
  nextEventId: number;
  startsAt: number;
  subtitle: string;
  subtitled: boolean;
  title: string;
  widescreen: boolean;
  dvrUuid?: string;
  dvrState?: string;
};

export type EpgChannel = {
  channelId: string;
  channelName: string;
  channelNumber: number;
  piconId: number;
  events: EpgEvent[];
};

export type Channel = {
  id: string;
  enabled: boolean;
  name: string;
  number: number;
  piconId: number;
};
