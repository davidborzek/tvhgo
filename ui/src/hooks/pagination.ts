import { useState } from 'react';
import { SetURLSearchParams } from 'react-router-dom';
import { URLSearchParams } from 'url';

export const usePagination = (
  initialLimit: number = 50,
  searchParams: URLSearchParams,
  setSearchParams: SetURLSearchParams
) => {
  const [limit, setLimit] = useState(initialLimit);

  const getOffset = () => {
    const offset = searchParams.get('offset');
    return offset ? parseInt(offset) : 0;
  };

  const setOffset = (value: number) => {
    setSearchParams((prev) => {
      prev.set('offset', `${value}`);
      return prev;
    });
  };

  const nextPage = () => {
    setOffset(getOffset() + limit);
  };

  const previousPage = () => {
    setOffset(getOffset() - limit);
  };

  const firstPage = () => {
    setOffset(0);
  };

  const lastPage = (total: number) => {
    setOffset(Math.floor(total / limit) * limit);
  };

  return {
    limit,
    nextPage,
    previousPage,
    getOffset,
    lastPage,
    firstPage,
    setLimit,
  };
};
