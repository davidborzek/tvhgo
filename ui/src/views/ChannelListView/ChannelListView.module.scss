.container {
  overflow: auto;
  padding: 2rem;
}

.channelList {
  max-width: 1400px;
  margin: 0 auto;
  margin-bottom: 2rem;
}
