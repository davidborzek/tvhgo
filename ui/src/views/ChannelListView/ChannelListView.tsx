import { useEffect, useRef } from 'react';
import { useFetchEvents } from '../../hooks/epg';

import styles from './ChannelListView.module.scss';
import { useTranslation } from 'react-i18next';
import ChannelListItem from '../../components/ChannelListItem/ChannelListItem';
import Error from '../../components/Error/Error';
import { useNavigate, useSearchParams } from 'react-router-dom';
import { usePagination } from '../../hooks/pagination';
import PaginationControls from '../../components/PaginationControls/PaginationControls';

const defaultLimit = 50;

function ChannelListView() {
  const ref = useRef<HTMLDivElement>(null);
  const { t } = useTranslation();
  const navigate = useNavigate();
  const [searchParams, setSearchParams] = useSearchParams();

  const { limit, nextPage, previousPage, getOffset, firstPage, lastPage } =
    usePagination(defaultLimit, searchParams, setSearchParams);

  const { events, total, error } = useFetchEvents({
    nowPlaying: true,
    limit,
    offset: getOffset(),
    sort_key: 'channelNumber',
    sort_dir: 'asc',
  });

  useEffect(() => {
    const scrollPos = searchParams.get('pos');

    if (events.length > 0) {
      ref.current?.scrollTo(0, parseInt(scrollPos || '0'));
    }
  }, [events, searchParams]);

  const renderChannels = () => {
    return events.map((event) => {
      return (
        <ChannelListItem
          key={event.id}
          event={event}
          onClick={(id) => {
            if (ref.current?.scrollTop !== undefined) {
              const pos = Math.floor(ref.current?.scrollTop);
              setSearchParams((prev) => {
                prev.set('pos', `${Math.floor(pos)}`);
                return prev;
              });
            }

            navigate(`/channels/${id}`);
          }}
        />
      );
    });
  };

  if (error) {
    return <Error message={error} />;
  }

  return (
    <div ref={ref} className={styles.container}>
      <div className={styles.channelList}>
        <h1>{t('channels')}</h1>
        {renderChannels()}
      </div>
      <PaginationControls
        onNextPage={nextPage}
        onPreviousPage={previousPage}
        onFirstPage={firstPage}
        onLastPage={() => lastPage(total)}
        scrollTop={() => {
          setSearchParams((prev) => {
            prev.delete('pos');
            return prev;
          });
          ref.current?.scrollTo(0, 0);
        }}
        limit={limit}
        offset={getOffset()}
        total={total}
      />
    </div>
  );
}

export default ChannelListView;
