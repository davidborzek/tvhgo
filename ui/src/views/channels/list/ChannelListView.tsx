import { EpgEvent, ListResponse } from '@/clients/api/api.types';
import {
  LoaderFunctionArgs,
  useLoaderData,
  useNavigate,
  useSearchParams,
} from 'react-router-dom';

import ChannelListItem from '@/components/channels/listItem/ChannelListItem';
import ChannelViewer from '@/components/channels/viewer/ChannelViewer';
import EmptyState from '@/components/common/emptyState/EmptyState';
import PaginationControls from '@/components/common/paginationControls/PaginationControls';
import { getEpgEvents } from '@/clients/api/api';
import styles from './ChannelListView.module.scss';
import { usePagination } from '@/hooks/pagination';
import { useState } from 'react';
import { useTranslation } from 'react-i18next';

const defaultLimit = 50;

export async function loader({ request }: LoaderFunctionArgs) {
  const query = new URL(request.url).searchParams;

  return getEpgEvents({
    nowPlaying: true,
    limit: defaultLimit,
    offset: parseInt(query.get('offset') || '0') || 0,
    // eslint-disable-next-line camelcase
    sort_key: 'channelNumber',
    // eslint-disable-next-line camelcase
    sort_dir: 'asc',
  });
}

export function Component() {
  const { t } = useTranslation();
  const navigate = useNavigate();
  const [searchParams, setSearchParams] = useSearchParams();
  const [selectedEvent, setSelectedEvent] = useState<EpgEvent | null>(null);
  const [isViewerOpen, setIsViewerOpen] = useState(false);

  const { limit, nextPage, previousPage, getOffset, firstPage, lastPage } =
    usePagination(defaultLimit, searchParams, setSearchParams);

  const { entries, total } = useLoaderData() as ListResponse<EpgEvent>;

  const handleWatch = (event: EpgEvent) => {
    setSelectedEvent(event);
    setIsViewerOpen(true);
  };

  const handleCloseViewer = () => {
    setIsViewerOpen(false);
    setSelectedEvent(null);
  };

  if (entries.length === 0) {
    return (
      <EmptyState title={t('no_channels')} subtitle={t('no_channels_info')} />
    );
  }

  const renderChannels = () => {
    return entries.map((event) => {
      return (
        <ChannelListItem
          key={event.channelId}
          event={event}
          onClick={(id) => {
            navigate(`/channels/${id}`);
          }}
          onWatch={handleWatch}
        />
      );
    });
  };

  return (
    <div className={styles.container}>
      <div className={styles.channelList}>{renderChannels()}</div>
      <PaginationControls
        onNextPage={nextPage}
        onPreviousPage={previousPage}
        onFirstPage={firstPage}
        onLastPage={() => lastPage(total)}
        limit={limit}
        offset={getOffset()}
        total={total}
      />
      {selectedEvent && (
        <ChannelViewer
          event={selectedEvent}
          isOpen={isViewerOpen}
          onClose={handleCloseViewer}
        />
      )}
    </div>
  );
}

Component.displayName = 'ChannelListView';
