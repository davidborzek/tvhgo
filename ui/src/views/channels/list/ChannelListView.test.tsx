import { EpgEvent } from '@/clients/api/api.types';
import { cleanup, render } from '@testing-library/react';
import { useLoaderData, useNavigate, useSearchParams } from 'react-router-dom';
import { afterEach, beforeEach, expect, test, vi, describe } from 'vitest';
import { Component as ChannelListView } from './ChannelListView';
import userEvent from '@testing-library/user-event';
import { TestIds } from '@/__test__/ids';
import { usePagination } from '@/hooks/pagination';

vi.mock('react-router-dom');
vi.mock('@/hooks/pagination');

const buildChannel = (id: string, name: string): EpgEvent => {
  return {
    audioDesc: true,
    channelId: id,
    channelName: name,
    channelNumber: 1,
    description: `${name} - Description`,
    endsAt: 0,
    hd: true,
    id: 1,
    nextEventId: 2,
    piconId: 1,
    startsAt: 0,
    subtitle: `${name} - Subtitle`,
    subtitled: true,
    title: `${name} - Title`,
    widescreen: true,
  };
};

const buildChannels = (count: number): EpgEvent[] => {
  return [...new Array(count)].map((_, i) =>
    buildChannel(`id_${i}`, `Channel_${i}`)
  );
};

const navigateMock = vi.fn();
const setQueryParamsMock = vi.fn();
const nextPageMock = vi.fn();
const previousPageMock = vi.fn();
const getOffsetMock = vi.fn();
const firstPageMock = vi.fn();
const lastPageMock = vi.fn();

beforeEach(() => {
  vi.mocked(useSearchParams).mockReturnValue([
    new URLSearchParams(),
    setQueryParamsMock,
  ]);
  vi.mocked(useNavigate).mockReturnValue(navigateMock);
  vi.mocked(usePagination).mockReturnValue({
    firstPage: firstPageMock,
    getOffset: getOffsetMock,
    lastPage: lastPageMock,
    limit: 50,
    nextPage: nextPageMock,
    previousPage: previousPageMock,
    setLimit: vi.fn(),
  });
});

afterEach(() => {
  vi.resetAllMocks();
  cleanup();
});

test('render empty state', () => {
  vi.mocked(useLoaderData).mockReturnValue({
    entries: [],
    offset: 0,
    total: 0,
  });
  getOffsetMock.mockReturnValue(0);

  const document = render(<ChannelListView />);

  expect(document.asFragment()).toMatchSnapshot();
});

test('render with one page', () => {
  const channels = buildChannels(5);
  vi.mocked(useLoaderData).mockReturnValue({
    entries: channels,
    offset: 0,
    total: channels.length,
  });
  getOffsetMock.mockReturnValue(0);

  const document = render(<ChannelListView />);

  expect(document.asFragment()).toMatchSnapshot();
});

test('render with multiple pages', () => {
  const channels = buildChannels(5);
  vi.mocked(useLoaderData).mockReturnValue({
    entries: channels,
    offset: 0,
    total: 100,
  });
  getOffsetMock.mockReturnValue(0);

  const document = render(<ChannelListView />);

  expect(document.asFragment()).toMatchSnapshot();
});

describe('pagination', () => {
  test('go next page', async () => {
    const channels = buildChannels(5);
    vi.mocked(useLoaderData).mockReturnValue({
      entries: channels,
      offset: 0,
      total: 200,
    });
    getOffsetMock.mockReturnValue(0);

    const document = render(<ChannelListView />);

    await userEvent.click(document.getByTestId(TestIds.PAGINATION_NEXT_PAGE));
    expect(nextPageMock).toHaveBeenCalled();
  });

  test('go previous', async () => {
    const channels = buildChannels(5);
    vi.mocked(useLoaderData).mockReturnValue({
      entries: channels,
      offset: 50,
      total: 200,
    });
    getOffsetMock.mockReturnValue(50);

    const document = render(<ChannelListView />);

    await userEvent.click(
      document.getByTestId(TestIds.PAGINATION_PREVIOUS_PAGE)
    );
    expect(previousPageMock).toHaveBeenCalled();
  });

  test('go last', async () => {
    const channels = buildChannels(5);
    vi.mocked(useLoaderData).mockReturnValue({
      entries: channels,
      offset: 0,
      total: 200,
    });
    getOffsetMock.mockReturnValue(0);

    const document = render(<ChannelListView />);

    await userEvent.click(document.getByTestId(TestIds.PAGINATION_LAST_PAGE));
    expect(lastPageMock).toHaveBeenCalled();
  });

  test('go first', async () => {
    const channels = buildChannels(5);
    vi.mocked(useLoaderData).mockReturnValue({
      entries: channels,
      offset: 200,
      total: 200,
    });
    getOffsetMock.mockReturnValue(200);

    const document = render(<ChannelListView />);

    await userEvent.click(document.getByTestId(TestIds.PAGINATION_FIRST_PAGE));
    expect(firstPageMock).toHaveBeenCalled();
  });
});
