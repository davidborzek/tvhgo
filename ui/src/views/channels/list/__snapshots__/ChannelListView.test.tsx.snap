// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`render empty state 1`] = `
<DocumentFragment>
  <div
    class="_emptyState_2228c6"
  >
    <span
      class="_title_2228c6"
    >
      no_channels
    </span>
    <span
      class="_subtitle_2228c6"
    >
      no_channels_info
    </span>
  </div>
</DocumentFragment>
`;

exports[`render with multiple pages 1`] = `
<DocumentFragment>
  <div
    class="_container_0e2490"
  >
    <div
      class="_channelList_0e2490"
    >
      <div
        class="_channel_8f5b3b"
      >
        <div
          class="_piconContainer_8f5b3b"
        >
          <div
            class="_picon_8f5b3b"
          >
            <div
              style="width: 100%; height: 100%;"
            />
            <img
              alt=""
              class="_picon_8f5b3b"
              src="/api/picon/1"
              style="display: none;"
              title="Channel_0"
            />
          </div>
        </div>
        <div
          class="_event_8f5b3b"
        >
          <span
            class="_channelName_8f5b3b"
          >
            Channel_0
          </span>
          <span
            class="_eventTitle_8f5b3b"
          >
            Channel_0 - Title
          </span>
          <span
            class="_eventTitle_8f5b3b"
          >
            event_time
          </span>
        </div>
      </div>
      <div
        class="_channel_8f5b3b"
      >
        <div
          class="_piconContainer_8f5b3b"
        >
          <div
            class="_picon_8f5b3b"
          >
            <div
              style="width: 100%; height: 100%;"
            />
            <img
              alt=""
              class="_picon_8f5b3b"
              src="/api/picon/1"
              style="display: none;"
              title="Channel_1"
            />
          </div>
        </div>
        <div
          class="_event_8f5b3b"
        >
          <span
            class="_channelName_8f5b3b"
          >
            Channel_1
          </span>
          <span
            class="_eventTitle_8f5b3b"
          >
            Channel_1 - Title
          </span>
          <span
            class="_eventTitle_8f5b3b"
          >
            event_time
          </span>
        </div>
      </div>
      <div
        class="_channel_8f5b3b"
      >
        <div
          class="_piconContainer_8f5b3b"
        >
          <div
            class="_picon_8f5b3b"
          >
            <div
              style="width: 100%; height: 100%;"
            />
            <img
              alt=""
              class="_picon_8f5b3b"
              src="/api/picon/1"
              style="display: none;"
              title="Channel_2"
            />
          </div>
        </div>
        <div
          class="_event_8f5b3b"
        >
          <span
            class="_channelName_8f5b3b"
          >
            Channel_2
          </span>
          <span
            class="_eventTitle_8f5b3b"
          >
            Channel_2 - Title
          </span>
          <span
            class="_eventTitle_8f5b3b"
          >
            event_time
          </span>
        </div>
      </div>
      <div
        class="_channel_8f5b3b"
      >
        <div
          class="_piconContainer_8f5b3b"
        >
          <div
            class="_picon_8f5b3b"
          >
            <div
              style="width: 100%; height: 100%;"
            />
            <img
              alt=""
              class="_picon_8f5b3b"
              src="/api/picon/1"
              style="display: none;"
              title="Channel_3"
            />
          </div>
        </div>
        <div
          class="_event_8f5b3b"
        >
          <span
            class="_channelName_8f5b3b"
          >
            Channel_3
          </span>
          <span
            class="_eventTitle_8f5b3b"
          >
            Channel_3 - Title
          </span>
          <span
            class="_eventTitle_8f5b3b"
          >
            event_time
          </span>
        </div>
      </div>
      <div
        class="_channel_8f5b3b"
      >
        <div
          class="_piconContainer_8f5b3b"
        >
          <div
            class="_picon_8f5b3b"
          >
            <div
              style="width: 100%; height: 100%;"
            />
            <img
              alt=""
              class="_picon_8f5b3b"
              src="/api/picon/1"
              style="display: none;"
              title="Channel_4"
            />
          </div>
        </div>
        <div
          class="_event_8f5b3b"
        >
          <span
            class="_channelName_8f5b3b"
          >
            Channel_4
          </span>
          <span
            class="_eventTitle_8f5b3b"
          >
            Channel_4 - Title
          </span>
          <span
            class="_eventTitle_8f5b3b"
          >
            event_time
          </span>
        </div>
      </div>
    </div>
    <div
      class="_controls_f3f6fd"
    >
      <button
        class="_button_09f37a _disabled_09f37a"
        data-testid="pagination-first-page-btn"
        disabled=""
      >
        &lt;&lt;
      </button>
      <button
        class="_button_09f37a _disabled_09f37a"
        data-testid="pagination-previous-page-btn"
        disabled=""
      >
        &lt;
      </button>
      <span
        class="_page_f3f6fd"
      >
        pagination_info
      </span>
      <button
        class="_button_09f37a"
        data-testid="pagination-next-page-btn"
      >
        &gt;
      </button>
      <button
        class="_button_09f37a"
        data-testid="pagination-last-page-btn"
      >
        &gt;&gt;
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`render with one page 1`] = `
<DocumentFragment>
  <div
    class="_container_0e2490"
  >
    <div
      class="_channelList_0e2490"
    >
      <div
        class="_channel_8f5b3b"
      >
        <div
          class="_piconContainer_8f5b3b"
        >
          <div
            class="_picon_8f5b3b"
          >
            <div
              style="width: 100%; height: 100%;"
            />
            <img
              alt=""
              class="_picon_8f5b3b"
              src="/api/picon/1"
              style="display: none;"
              title="Channel_0"
            />
          </div>
        </div>
        <div
          class="_event_8f5b3b"
        >
          <span
            class="_channelName_8f5b3b"
          >
            Channel_0
          </span>
          <span
            class="_eventTitle_8f5b3b"
          >
            Channel_0 - Title
          </span>
          <span
            class="_eventTitle_8f5b3b"
          >
            event_time
          </span>
        </div>
      </div>
      <div
        class="_channel_8f5b3b"
      >
        <div
          class="_piconContainer_8f5b3b"
        >
          <div
            class="_picon_8f5b3b"
          >
            <div
              style="width: 100%; height: 100%;"
            />
            <img
              alt=""
              class="_picon_8f5b3b"
              src="/api/picon/1"
              style="display: none;"
              title="Channel_1"
            />
          </div>
        </div>
        <div
          class="_event_8f5b3b"
        >
          <span
            class="_channelName_8f5b3b"
          >
            Channel_1
          </span>
          <span
            class="_eventTitle_8f5b3b"
          >
            Channel_1 - Title
          </span>
          <span
            class="_eventTitle_8f5b3b"
          >
            event_time
          </span>
        </div>
      </div>
      <div
        class="_channel_8f5b3b"
      >
        <div
          class="_piconContainer_8f5b3b"
        >
          <div
            class="_picon_8f5b3b"
          >
            <div
              style="width: 100%; height: 100%;"
            />
            <img
              alt=""
              class="_picon_8f5b3b"
              src="/api/picon/1"
              style="display: none;"
              title="Channel_2"
            />
          </div>
        </div>
        <div
          class="_event_8f5b3b"
        >
          <span
            class="_channelName_8f5b3b"
          >
            Channel_2
          </span>
          <span
            class="_eventTitle_8f5b3b"
          >
            Channel_2 - Title
          </span>
          <span
            class="_eventTitle_8f5b3b"
          >
            event_time
          </span>
        </div>
      </div>
      <div
        class="_channel_8f5b3b"
      >
        <div
          class="_piconContainer_8f5b3b"
        >
          <div
            class="_picon_8f5b3b"
          >
            <div
              style="width: 100%; height: 100%;"
            />
            <img
              alt=""
              class="_picon_8f5b3b"
              src="/api/picon/1"
              style="display: none;"
              title="Channel_3"
            />
          </div>
        </div>
        <div
          class="_event_8f5b3b"
        >
          <span
            class="_channelName_8f5b3b"
          >
            Channel_3
          </span>
          <span
            class="_eventTitle_8f5b3b"
          >
            Channel_3 - Title
          </span>
          <span
            class="_eventTitle_8f5b3b"
          >
            event_time
          </span>
        </div>
      </div>
      <div
        class="_channel_8f5b3b"
      >
        <div
          class="_piconContainer_8f5b3b"
        >
          <div
            class="_picon_8f5b3b"
          >
            <div
              style="width: 100%; height: 100%;"
            />
            <img
              alt=""
              class="_picon_8f5b3b"
              src="/api/picon/1"
              style="display: none;"
              title="Channel_4"
            />
          </div>
        </div>
        <div
          class="_event_8f5b3b"
        >
          <span
            class="_channelName_8f5b3b"
          >
            Channel_4
          </span>
          <span
            class="_eventTitle_8f5b3b"
          >
            Channel_4 - Title
          </span>
          <span
            class="_eventTitle_8f5b3b"
          >
            event_time
          </span>
        </div>
      </div>
    </div>
    <div
      class="_controls_f3f6fd"
    >
      <button
        class="_button_09f37a _disabled_09f37a"
        data-testid="pagination-first-page-btn"
        disabled=""
      >
        &lt;&lt;
      </button>
      <button
        class="_button_09f37a _disabled_09f37a"
        data-testid="pagination-previous-page-btn"
        disabled=""
      >
        &lt;
      </button>
      <span
        class="_page_f3f6fd"
      >
        pagination_info
      </span>
      <button
        class="_button_09f37a _disabled_09f37a"
        data-testid="pagination-next-page-btn"
        disabled=""
      >
        &gt;
      </button>
      <button
        class="_button_09f37a _disabled_09f37a"
        data-testid="pagination-last-page-btn"
        disabled=""
      >
        &gt;&gt;
      </button>
    </div>
  </div>
</DocumentFragment>
`;
